h/api/models/test/document_test.py

```diff
@@ -119,6 +122,416 @@ def test_document_find_or_create_by_uris_no_results():
+        document=mock_document(),
+        created=datetime.datetime.now(),
+        updated=datetime.datetime.now())
+
+    assert not DocumentURI.called
+    assert not db.add.called
+
+
+@create_or_update_document_uri_fixtures
+def test_create_or_update_document_uri_logs_warning_if_document_ids_differ(
+        log,
+        DocumentURI):
+    """
+    It should log a warning on Document objects mismatch.
+
+    If there's an existing DocumentURI and its .document property is different
+    to the given document it shoulg log a warning.
+
+    """
+    # existing_document_uri.document will not be equal to the given document'
+    existing_document_uri = mock.Mock(document=mock_document())
+    DocumentURI.query.filter.return_value.first.return_value = existing_document_uri
+
+    document.create_or_update_document_uri(
+        db=mock_db(),
+        claimant='http://example.com/example_claimant.html',
+        uri='http://example.com/example_uri.html',
+        type='self-claim',
+        content_type=None,
+        document=mock_document(),
+        created=datetime.datetime.now(),
+        updated=datetime.datetime.now())
+
+    assert log.warn.call_count == 1
+
+
+@create_or_update_document_uri_fixtures
+def test_create_or_update_document_uri_updates_updated_time(DocumentURI):
+    # existing_document_uri has an older .updated time than than the updated
+    # argument we will pass to create_or_update_document_uri().
+    now = datetime.datetime.now()
+    yesterday = now - datetime.timedelta(days=1)
+    existing_document_uri = mock.Mock(updated=yesterday)
+
+    DocumentURI.query.filter.return_value.first.return_value = (
+        existing_document_uri)
+
+    document.create_or_update_document_uri(
+        db=mock_db(),
+        claimant='http://example.com/example_claimant.html',
+        uri='http://example.com/example_uri.html',
+        type='self-claim',
+        content_type=None,
+        document=mock_document(),
+        created=now - datetime.timedelta(days=3),
+        updated=now)
+
+    assert existing_document_uri.updated == now
+
+
+@pytest.mark.usefixtures('DocumentMeta',
+                         'log')
+class TestCreateOrUpdateDocumentMeta(object):
+
+    def test_it_calls_filter(self, DocumentMeta):
+        document.create_or_update_document_meta(
+            db=mock_db(),
+            claimant='http://example.com/claimant',
+            claimant_normalized='http://example.com/claimant_normalized',
+            type='title',
+            value='Example Page',
+            document=mock.Mock(),
+            created=yesterday(),
+            updated=now(),
+        )
+
+        # FIXME: We need to assert that this is called with the right
+        # arguments, but that's very awkward to do with the way sqlalchemy
+        # querying works.' Move this functionality onto the model class itself
+        # where the tests can use the actual database.
+        assert DocumentMeta.query.filter.call_count == 1
```

Comment:
Similar to what I mentioned above, why not just create a document with a document meta and then check that it doesn't initialize a new one?